subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/metadata/GradleModuleMetadataCompatibilityConverter.java

```diff
@@ -40,6 +42,28 @@ public GradleModuleMetadataCompatibilityConverter(ImmutableAttributesFactory att
     }
 
     public void process(MutableModuleComponentResolveMetadata metaDataFromResource) {
+        handleAttributeCompatibility(metaDataFromResource);
+        handleMavenSnapshotCompatibility(metaDataFromResource);
+    }
+
+    private void handleMavenSnapshotCompatibility(MutableModuleComponentResolveMetadata metaDataFromResource) {
+        if (metaDataFromResource.getId() instanceof MavenUniqueSnapshotComponentIdentifier) {
+            // Action needed only for Maven unique snapshots
+            // Verify that the URL of the artifacts properly references the unique version and not -SNAPSHOT
+            MavenUniqueSnapshotComponentIdentifier uniqueIdentifier = (MavenUniqueSnapshotComponentIdentifier) metaDataFromResource.getId();
+            for (MutableComponentVariant mutableVariant : metaDataFromResource.getMutableVariants()) {
+                for (ComponentVariant.File file : mutableVariant.getFiles()) {
+                    if (file.getUri().contains("SNAPSHOT")) {
+                        mutableVariant.removeFile(file);
+                        mutableVariant.addFile(file.getName(), file.getUri().replace("SNAPSHOT", uniqueIdentifier.getTimestamp()));
```

Comment:
Yes, though now that I think of it, I would only need a copy if and only if there are invalid entries in there. Which is bound to be less and less often the case.
Will refactor to do the copy here since the type is internal and so returning the real list should not be an issue.